### ISSUE TITLE
New Decoder/Encoder Structure

### ISSUE THREAD

**Issue by ilan-gold** [gao-lab/GLUE#14]
In the [docs](https://scglue.readthedocs.io/en/latest/dev.html#register-custom-encoder-and-decoder) you mention registering custom decoders/encoders which is exactly what I want to do.  However, I'm not totally clear on how to achieve this since the functions referenced are internal.  Is there a way to modify them in the API?  Some `python` trick I am unaware of? Or is there some sort of `register` function for decoders/encoders like there is for datasets?

**Comment by Jeff1995** [comment]
Sorry for the inconvenience. There is currently no `register` function like this. You would have to install `scglue` in dev mode and change the package code. (Pull request is welcomed!)

A hacky workaround however is to overwrite the following two functions so that your custom encoder and decoder can be selected:

* [`scglue.models.scglue.select_encoder`](https://github.com/gao-lab/GLUE/blob/043042906fea9b878cfc2704f1a45b0aac74316d/scglue/models/scglue.py#L40)
* [`scglue.models.scglue.select_decoder`](https://github.com/gao-lab/GLUE/blob/043042906fea9b878cfc2704f1a45b0aac74316d/scglue/models/scglue.py#L61)

I think you might be able to do this by something like this:

```python
class MyEncoder:
    pass


class MyDecoder:
    pass


def select_encoder(prob_model):
    # Copy previous models here
    if prob_model == "my_prob_model":
        return MyEncoder


def select_decoder(prob_model):
    # Copy previous models here
    if prob_model == "my_prob_model":
        return MyDecoder


scglue.models.scglue.select_encoder = select_encoder
scglue.models.scglue.select_decoder = select_decoder
```

I agree that a `register` function would be much better. It will be supported in the next release. I'll let you know when it's ready!

**Comment by ilan-gold** [comment]
Much appreciated!

**Comment by Jeff1995** [comment]
A new [`register_prob_model`](https://scglue.readthedocs.io/en/v0.2.2/api/scglue.models.scglue.register_prob_model.html) function is now available in v0.2.2! 😀

**Comment by ilan-gold** [comment]
Excellent, gonna give it a whirl!
